File: Week15/LeetCode-0743-2.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        """
        This time we run the Bellmann-Ford algorithm
        """        
        maxDist = 100*100
        dist = [maxDist]*(n+1)
        dist[0] = 0      # There is no node with id=0. So, make its distance 0
        dist[k] = 0
        
        for ite in range(n-1):
            for u, v, w in times:
                dist[v] = min(dist[v], dist[u]+w) # Relax(u, v)
        
        maxVal = max(dist)
        return -1 if maxVal == maxDist else maxVal

File: Week15/LeetCode-0743-2.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        /// This time we run the Bellmann-Ford algorithm
        int maxDist = 100*100;
        vector<int> dist(n+1, maxDist);
        dist[k] = 0;
        
        for (int iter=0; iter<n-1; iter++){
            for (int i=0; i<times.size(); i++){
                int u = times[i][0];
                int v = times[i][1];
                int w = times[i][2];
                
                dist[v] = min(dist[v], dist[u]+w); // Relax(u, v)
            } // end-for         
        } // end-for
        
        int maxVal = *std::max_element(dist.begin()+1, dist.end());        
        return maxVal == maxDist? -1: maxVal;
    } // end-for
};